admin/code/CampaignAdmin.php

```diff
@@ -465,19 +443,50 @@ public function publishCampaign(SS_HTTPRequest $request) {
 	}
 
 	/**
+	 * Url handler for edit form
+	 *
+	 * @param SS_HTTPRequest $request
+	 * @return Form
+	 */
+	public function DetailEditForm($request) {
+		// Get ID either from posted back value, or url parameter
+		$id = $request->param('ID') ?: $request->postVar('ID');
+		return $this->getDetailEditForm($id);
+	}
+
+	/**
 	 * @todo Use GridFieldDetailForm once it can handle structured data and form schemas
 	 *
+	 * @param int $id
 	 * @return Form
 	 */
-	public function getDetailEditForm() {
-		return Form::create(
+	public function getDetailEditForm($id) {
+		// Get record-specific fields
+		$record = null;
+		if($id) {
+			$record = ChangeSet::get()->byId($id);
+		}
+		if(!$record) {
+			$record = ChangeSet::singleton();
+		}
+		$fields = $record->getCMSFields();
+
+		// Add standard fields
+		$fields->push(HiddenField::create('ID'));
+		$form = Form::create(
 			$this,
 			'DetailEditForm',
-			ChangeSet::singleton()->getCMSFields(),
+			$fields,
 			FieldList::create(
 				FormAction::create('save', 'Save')
 			)
 		);
+		// Configure form to respond to validation errors with form schema
+		// if requested via react.
+		$form->setValidationResponseCallback(function() use ($form) {
+			return $this->getSchemaResponse($form);
```

Comment:
`Campaign` is not enforcing `Title` as a required field at the moment - which is fine for this PR, but we should track this in a Github issue.